### DISCUSSION TITLE
Support NGettext

### DISCUSSION THREAD

**Discussion by Slesa** [AvaloniaUI/Avalonia#9723] [General]
It is great that i18n ressources can be loaded from XAML. On the other hand: it would be nice to use something like gettext with its po files, as it already has PO Editors and other tools.

For WPF, the solution was NGettext-WPF so far, which is a XAML-based approach and it works like:
```xaml
        <TextBlock  Text="{wpf:Gettext This is a translated text}" />
```

In code behind, text can be translated, too:
```cs
Translation.GetParticularString("CodeBehind", "Another translated text");
```

NGettext-WPF still uses .NET 4.6, so I'd like to convert the project files to .NET6 first.
But also, I'd like to add NGettext-Avalonia, using the same mechanism. My question here is: would that be a part of Avalonia itself, or should it be a standalone Nuget proejct?

**Comment by grokys** [comment]
Personally, I think a standalone project - I'm not sure we have the manpower or expertise to maintain our own localization solution and it would also probably discourage people like yourselves from working on alternatives too.

**Comment by Slesa** [comment]
The only thing so far: when changing the language, the current content of a combobox is not updated.
Here is a first draft. 
https://github.com/Slesa/ngettext-avalonia

I can't see many more alternatives than
- Load from resx/xaml file
- Use .PO files
- Read from a database

**Reply by maxkatz6** [reply]
> I can't see many more alternatives than

- Keeping resources in the resource dictionaries.
- Creating own markup extension (just like NGetText, but you are not limited with single option)

**Reply by maxkatz6** [reply]
> the current content of a combobox is not updated.

Checked, not sure how it should work. You don't have any custom item template set with this extension inside.
So, it's basically the same as `Enum.ToString()`.

**Reply by maxkatz6** [reply]
Also checked the code to see how it works.
Not sure how WPF handles it, but in Avalonia implementation you have a potential problem.
You call HandleCultureChanged on the extension which then sets property value again with SetValue. Basically, bypassing value store infra. 

It will break, when you use extension in styles - `<Setter Value="{avalonia:Gettext English}" />` - which will create a value with "Style" priority. Same property might have multiple style priorities (multiple styles applied on the property), and it has a lower priority comparing to the Local value.
But when you call `object.SetValue()` without specifying a priority, it will set it as a local value, overring everything else, basically ruining hierarchy of value store priorities.

Alternative, implementation, safer, would be to return a binding instance from the extension. Like we do in DynamicResource https://github.com/AvaloniaUI/Avalonia/blob/master/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/DynamicResourceExtension.cs#L41

**Comment by alexandrehtrb** [comment]
I use this solution for multi-language support in Avalonia, it works very well:

https://www.sakya.it/wordpress/avalonia-ui-framework-localization/

https://github.com/sakya/AvaloniaLocalizationExample

**Comment by Slesa** [comment]
Well, the reason why I prefer the ngettext stuff is the use of the PoEdit and string exttraction tools instead of fiddling with XML/XAML, databases or json. When using Avalonia for business applications, there is always a translator which has no clue how to use git or a text editor. 

Nevertheless, it's a valid solution. Maybe there is an example missing how to deal with placeholders, e.g. 
"Hello %s, your flight will at %d:%d", user, hour, minute
and what happens if you need to switch the order.
